File: src/main/java/komu/blunt/asm/opcodes/OpCopyRegister.java
package komu.blunt.asm.opcodes;

import komu.blunt.asm.Register;
import komu.blunt.asm.VM;

import static com.google.common.base.Preconditions.checkNotNull;

public class OpCopyRegister extends OpCode {
    private final Register target;
    private final Register source;


    public OpCopyRegister(Register target, Register source) {
        this.target = checkNotNull(target);
        this.source = checkNotNull(source);
    }

    @Override
    public void execute(VM vm) {
        vm.set(target, vm.get(source));
    }

    @Override
    public String toString() {
        return String.format("(copy %s %s)", target, source);
    }
}

File: src/main/kotlin/komu/blunt/asm/opcodes/OpCopyRegister.kt
package komu.blunt.asm.opcodes

import komu.blunt.asm.Register
import komu.blunt.asm.VM

class OpCopyRegister(private val target: Register, private val source: Register) : OpCode() {

    override fun execute(vm: VM) {
        vm[target] = vm[source]
    }

    override fun modifies(register: Register) = register == target
    override fun toString() = "copy $target $source"
}
